File: Wikipedia/Code/WMFSearchLanguagesBarViewController.m
#import "WMFSearchLanguagesBarViewController.h"
#import "UIFont+WMFStyle.h"
#import "WMFLanguagesViewController.h"

@interface WMFSearchLanguagesBarViewController () <WMFLanguagesViewControllerDelegate>

@property (strong, nonatomic) IBOutlet NSLayoutConstraint *heightContraint;
@property (strong, nonatomic) IBOutlet UIButton *languageOneButton;
@property (strong, nonatomic) IBOutlet UIButton *languageTwoButton;
@property (strong, nonatomic) IBOutlet UIButton *languageThreeButton;
@property (strong, nonatomic) IBOutlet UIButton *otherLanguagesButton;
@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *languageButtons;
@property (strong, nonatomic) MWKLanguageLink *previousFirstLanguage;
@property (strong, nonatomic) MWKLanguageLink* currentlySelectedSearchLanguage;
@property (nonatomic) BOOL hidden;

@end

@implementation WMFSearchLanguagesBarViewController

@synthesize currentlySelectedSearchLanguage = _currentlySelectedSearchLanguage;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.languageButtons enumerateObjectsUsingBlock:^(UIButton *_Nonnull button, NSUInteger idx, BOOL *_Nonnull stop) {
        button.tintColor = [UIColor wmf_blueTintColor];
    }];
    [self.otherLanguagesButton setBackgroundImage:[UIImage wmf_imageFromColor:[UIColor whiteColor]] forState:UIControlStateNormal];
    [self.otherLanguagesButton setBackgroundImage:[UIImage wmf_imageFromColor:[UIColor colorWithWhite:0.9 alpha:1]] forState:UIControlStateHighlighted];
    [self.otherLanguagesButton setTitle:MWLocalizedString(@"main-menu-title", nil) forState:UIControlStateNormal];
    self.otherLanguagesButton.titleLabel.font = [UIFont wmf_subtitle];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self updateLanguageBarLanguageButtons];
    self.hidden = ![[NSUserDefaults wmf_userDefaults] wmf_showSearchLanguageBar];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged];
    self.previousFirstLanguage = [self.languageBarLanguages firstObject];
}

- (void)selectFirstLanguageIfNoneSelectedOrIfFirstLanguageHasChanged {
    if([self isEveryButtonUnselected] || [self isFirstLanguageDifferentFromLastTime]){
        [self setLanguageWithSender:self.languageButtons.firstObject];
    }
}

- (BOOL)isEveryButtonUnselected{
    for(UIButton *button in self.languageButtons){
        if(button.selected){
            return NO;
        }
    }
    return YES;
}

- (BOOL)isFirstLanguageDifferentFromLastTime {
    return ![self.previousFirstLanguage isEqualToLanguageLink:[self.languageBarLanguages firstObject]];
}

- (void)setHidden:(BOOL)hidden {
    if(hidden){
        self.heightContraint.constant = 0;
        self.view.hidden = YES;
    }else{
        self.heightContraint.constant = 44;
        self.view.hidden = NO;
    }
    _hidden = hidden;
}

- (NSArray<MWKLanguageLink *> *)languageBarLanguages {
    return [[MWKLanguageLinkController sharedInstance].preferredLanguages wmf_arrayByTrimmingToLength:3];
}

- (void)updateLanguageBarLanguageButtons {
    [self.languageBarLanguages enumerateObjectsUsingBlock:^(MWKLanguageLink *_Nonnull language, NSUInteger idx, BOOL *_Nonnull stop) {
        if (idx >= [self.languageButtons count]) {
            *stop = YES;
        }
        UIButton *button = self.languageButtons[idx];
        [button setTitle:[language localizedName] forState:UIControlStateNormal];
        [button setSelected:[[language siteURL] isEqual:[[self selectedLanguage] siteURL]]];
    }];
    
    [self.languageButtons enumerateObjectsUsingBlock:^(UIButton *_Nonnull obj, NSUInteger idx, BOOL *_Nonnull stop) {
        if (idx >= [self.languageBarLanguages count]) {
            obj.enabled = NO;
            obj.hidden = YES;
        } else {
            obj.enabled = YES;
            obj.hidden = NO;
        }
    }];
}

- (IBAction)setLanguageWithSender:(UIButton *)sender {
    NSUInteger index = [self.languageButtons indexOfObject:sender];
    NSAssert(index != NSNotFound, @"language button not found for language!");
    if (index != NSNotFound) {
        self.currentlySelectedSearchLanguage = [self.languageBarLanguages wmf_safeObjectAtIndex:index];
    }
}

- (void)setCurrentlySelectedSearchLanguage:(MWKLanguageLink *)currentlySelectedSearchLanguage {
    _currentlySelectedSearchLanguage = currentlySelectedSearchLanguage;
 
    [[NSUserDefaults wmf_userDefaults] wmf_setCurrentSearchLanguageDomain:currentlySelectedSearchLanguage.siteURL];
    [self.delegate searchLanguagesBarController:self didChangeCurrentlySelectedSearchLanguage:currentlySelectedSearchLanguage];
    [self updateLanguageBarLanguageButtons];
}

- (MWKLanguageLink *)selectedLanguage {
    NSURL *siteURL = [[NSUserDefaults wmf_userDefaults] wmf_currentSearchLanguageDomain];
    return siteURL ? [[MWKLanguageLinkController sharedInstance] languageForSiteURL:siteURL] : [self appLanguage];
}

- (nullable MWKLanguageLink *)appLanguage {
    MWKLanguageLink *language = [[MWKLanguageLinkController sharedInstance] appLanguage];
    NSAssert(language, @"No app language data found");
    return language;
}

- (IBAction)openLanguagePicker:(id)sender {
    WMFLanguagesViewController *languagesVC = [WMFPreferredLanguagesViewController preferredLanguagesViewController];
    languagesVC.delegate = self;
    [self presentViewController:[[UINavigationController alloc] initWithRootViewController:languagesVC] animated:YES completion:NULL];
}

- (void)languagesController:(WMFPreferredLanguagesViewController *)controller didUpdatePreferredLanguages:(NSArray<MWKLanguageLink *> *)languages {
    [self updateLanguageBarLanguageButtons];
}

@end

File: Wikipedia/Code/WMFSearchLanguagesBarViewController.swift
import Foundation

@objc protocol WMFSearchLanguagesBarViewControllerDelegate: class {
    func searchLanguagesBarViewController(controller: WMFSearchLanguagesBarViewController, didChangeCurrentlySelectedSearchLanguage language: MWKLanguageLink)
}

class WMFSearchLanguagesBarViewController: UIViewController, WMFPreferredLanguagesViewControllerDelegate, WMFLanguagesViewControllerDelegate {
    weak var delegate: WMFSearchLanguagesBarViewControllerDelegate?
    
    @IBOutlet private var languageButtons: [UIButton] = []
    @IBOutlet private var otherLanguagesButton: UIButton?
    @IBOutlet private var heightConstraint: NSLayoutConstraint?
    
    private var hidden: Bool = false {
        didSet {
            if(hidden){
                heightConstraint!.constant = 0
                view.hidden = true
            }else{
                heightConstraint!.constant = 44
                view.hidden = false
            }
        }
    }

    private(set) var currentlySelectedSearchLanguage: MWKLanguageLink? {
        get {
            if let siteURL = NSUserDefaults.wmf_userDefaults().wmf_currentSearchLanguageDomain(), selectedLanguage = MWKLanguageLinkController.sharedInstance().languageForSiteURL(siteURL) {
                return selectedLanguage
            }else{
                if let appLang:MWKLanguageLink? = MWKLanguageLinkController.sharedInstance().appLanguage {
                    self.currentlySelectedSearchLanguage = appLang
                    return appLang
                }else{
                    assert(false, "appLanguage should have been set at this point")
                    return nil
                }
            }
        }
        set {
            NSUserDefaults.wmf_userDefaults().wmf_setCurrentSearchLanguageDomain(newValue?.siteURL())
            delegate?.searchLanguagesBarViewController(self, didChangeCurrentlySelectedSearchLanguage: newValue!)
            updateLanguageBarLanguageButtons()
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        for button in languageButtons {
            button.tintColor = UIColor.wmf_blueTintColor()
        }
        otherLanguagesButton?.setBackgroundImage(UIImage.wmf_imageFromColor(UIColor.whiteColor()), forState: .Normal)
        otherLanguagesButton?.setBackgroundImage(UIImage.wmf_imageFromColor(UIColor(white: 0.9, alpha: 1.0)), forState: .Highlighted)
        otherLanguagesButton?.setTitle(localizedStringForKeyFallingBackOnEnglish("main-menu-title"), forState: .Normal)
        otherLanguagesButton?.titleLabel?.font = UIFont.wmf_subtitle()

        NSNotificationCenter.defaultCenter().addObserverForName(WMFAppLanguageDidChangeNotification, object: nil, queue: nil) { notification in
            guard let langController = notification.object, appLanguage = langController.appLanguage else {
                assert(false, "Could not extract app language from WMFAppLanguageDidChangeNotification")
                return
            }
            self.currentlySelectedSearchLanguage = appLanguage
        }

        NSNotificationCenter.defaultCenter().addObserverForName(WMFPreferredLanguagesDidChangeNotification, object: nil, queue: nil) { _ in
            if let selectedLang = self.currentlySelectedSearchLanguage {
                // The selected lang won't be in languageBarLanguages() if the user has dragged it down so it's not in top 3 langs...
                if(self.languageBarLanguages().indexOf(selectedLang) == nil){
                    // ...so select first lang if the selected lang has been moved down out of the top 3.
                    self.currentlySelectedSearchLanguage = self.languageBarLanguages().first
                    // Reminder: cannot use "reorderPreferredLanguage" for this (in "didUpdatePreferredLanguages:") because
                    // that would undo the dragging the user just did and would also not work for changes made from settings.
                }
            }
        }
    }

    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        updateLanguageBarLanguageButtons()
        hidden = !NSUserDefaults.wmf_userDefaults().wmf_showSearchLanguageBar()

        var selectedButtonCount = 0
        for button in languageButtons{
            if button.selected {
                selectedButtonCount += 1
            }
        }
//        assert(selectedButtonCount == 1, "One button should be selected by now")
    }
    
    private func languageBarLanguages() -> [MWKLanguageLink] {
        return Array(MWKLanguageLinkController.sharedInstance().preferredLanguages.prefix(3))
    }

    private func updateLanguageBarLanguageButtons(){
        for (index, language) in languageBarLanguages().enumerate() {
            if index >= languageButtons.count {
                break
            }
            let button = languageButtons[index]
            button.setTitle(language.localizedName, forState: .Normal)
            if let selectedLanguage = currentlySelectedSearchLanguage {
                button.selected = language.isEqualToLanguageLink(selectedLanguage)
            }else{
                assert(false, "selectedLanguage should have been set at this point")
                button.selected = false
            }
        }
        for(index, button) in languageButtons.enumerate(){
            if index >= languageBarLanguages().count {
                button.enabled = false
                button.hidden = true
            }else{
                button.enabled = true
                button.hidden = false
            }
        }
    }
    
    @IBAction private func setCurrentlySelectedLanguageToButtonLanguage(withSender sender: UIButton) {
        guard let buttonIndex = languageButtons.indexOf(sender) where languageBarLanguages().indices.contains(buttonIndex) else {
            assert(false, "Language button not found for language")
            return
        }
        currentlySelectedSearchLanguage = languageBarLanguages()[buttonIndex]
    }
    
    @IBAction private func openLanguagePicker() {
        let languagesVC = WMFPreferredLanguagesViewController.preferredLanguagesViewController()
        languagesVC.delegate = self
        presentViewController(UINavigationController.init(rootViewController: languagesVC), animated: true, completion: nil)
    }
    
    func languagesController(controller: WMFLanguagesViewController!, didSelectLanguage language: MWKLanguageLink!) {
        // If the selected language will not be displayed because we only display max 3 languages, move it to index 2
        if(languageBarLanguages().indexOf(language) == nil){
            MWKLanguageLinkController.sharedInstance().reorderPreferredLanguage(language, toIndex: 2)
        }
        
        currentlySelectedSearchLanguage = language
        controller.dismissViewControllerAnimated(true, completion: nil)
    }
}
